src/CSV.jl

```diff
@@ -364,141 +380,138 @@ function file(source,
+            tl_lastrefs = zeros(UInt64, ncols)
+            tl_tapes, tl_poslens = allocate(rowchunkguess, ncols, typecodes)
+            tl_datapos = ranges[i]
+            tl_len = ranges[i + 1] - (i != N)
+            tl_rows, tl_tapes = parsetape(Val(false), ignoreemptylines, ncols, gettypecodes(typemap), tl_tapes, tl_poslens, buf, tl_datapos, tl_len, limit, cmt, EMPTY_POSITIONS, pool, tl_refs, tl_lastrefs, rowchunkguess, typecodes, intsentinel, debug, options, true)
             debug && println("thread = $(Threads.threadid()): time for parsing: $(Base.time() - tt)")
-            rowsv[i] = ret[1]
-            tapesv[i] = ret[2]
-            refsv[i] = trefs
-            lastrefsv[i] = tlastrefs
+            perthreadrows[i] = tl_rows
+            perthreadtapes[i] = tl_tapes
+            perthreadrefs[i] = tl_refs
+            perthreadlastrefs[i] = tl_lastrefs
         end
 end # @static if VERSION >= v"1.3-DEV"
     end
-    rows = sum(rowsv)
-    tapes = Vector{UInt64}[Mmap.mmap(Vector{UInt64}, rows * 2) for i = 1:ncols]
     refs = Vector{Dict{String, UInt64}}(undef, ncols)
     lastrefs = zeros(UInt64, ncols)
-    rngs = Matrix{Vector{UInt64}}(undef, N, ncols)
-    recodes = falses(N, ncols)
     for i = 1:N
-        rs = refsv[i]
-        lrs = lastrefsv[i]
+        tlrefs = perthreadrefs[i]
+        tllastrefs = perthreadlastrefs[i]
         for col = 1:ncols
-            if isassigned(rs, col)
+            if typebits(typecodes[col]) == POOL && isassigned(tlrefs, col)
                 # merge refs and recode if necessary
                 if !isassigned(refs, col)
-                    refs[col] = rs[col]
-                    lastrefs[col] = lrs[col]
+                    refs[col] = tlrefs[col]
+                    lastrefs[col] = tllastrefs[col]
                 else
-                    rng = collect(UInt64(0):lrs[col])
+                    refrecodes = collect(UInt64(0):tllastrefs[col])
+                    colrefs = refs[col]
                     recode = false
-                    for (k, v) in rs[col]
-                        refvalue = get(refs[col], k, UInt64(0))
+                    for (k, v) in tlrefs[col]
+                        refvalue = get(colrefs, k, UInt64(0))
                         if refvalue != v
                             recode = true
                             if refvalue == 0
                                 refvalue = (lastrefs[col] += UInt64(1))
                             end
-                            refs[col][k] = refvalue
-                            rng[v + 1] = refvalue
+                            colrefs[k] = refvalue
+                            refrecodes[v + 1] = refvalue
                         end
                     end
-                    recodes[i, col] = recode
-                    rngs[i, col] = rng
-                end
-            end
-        end
-    end
-    @sync for j = 1:N
-@static if VERSION >= v"1.3-DEV"
```

Comment:
By using `LazyArrays`, it also made me realize we can clean up the ref-array syncing code in the multithreaded case. Instead of a two-step, complicated process, we now just do the single loop above and recode a pooled columns refs if needed.